### PR TITLE
Refactor helper pacakges and fix spec validation

### PR DESCRIPTION
**Pull request by AstroProfundis**
### What problem does this PR solve? 
The ports of monitoring agents are not checked with other existing clusters before deploying a new cluster.

### What is changed and how it works?
 - Refactor the package `prepare` and `clusterutils.Retry` to make it less possible of having import cycle.
 - Refactor the pre-deploy cross cluster conflict checks to involve ports of monitoring agents and make them easier to test
 - Add test cases for cross cluster conflict checks

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

### CODE REVIEW

File: pkg/cluster/spec/instance.go [hunk 1]
```diff
@@ -330,11 +330,20 @@ func (i *instance) GetPort() int {
 }
 
 func (i *instance) UsedPorts() []int {
-	return i.usedPorts
+	return append(
+		i.usedPorts,
+		i.topo.MonitoredOptions.NodeExporterPort,
```
Comment (lonng) [inline pkg/cluster/spec/instance.go:335]: These ports should be unique by the host not instance.

**Comment by lucklove** [comment]
Should we apply this logic in `CheckClusterDirConflict`?
We should check directory conflict, too.

File: pkg/cliutil/prepare/prepare.go [hunk 2]
```diff
@@ -216,6 +217,10 @@ func CheckClusterPortConflict(clusterSpec *meta.SpecManager, clusterName string,
 			return errors.Trace(err)
 		}
 
+		uniqueHosts := set.NewStringSet()
```
Comment (lonng) [inline pkg/cliutil/prepare/prepare.go:220]: Add some tests for this change.

**Comment by lonng** [comment]
> Should we apply this logic in `CheckClusterDirConflict`?
> We should check directory conflict, too.

I think we should.

**Comment by AstroProfundis** [comment]
There is already checks for dir conflicts with `hostDirAccessor`

File: pkg/cliutil/prepare/prepare.go [hunk 3]
```diff
@@ -231,13 +232,30 @@ func CheckClusterPortConflict(clusterSpec *spec.SpecManager, clusterName string,
 		}
 
 		metadata.GetTopology().IterInstance(func(inst spec.Instance) {
+			uniqueHosts := set.NewStringSet()
```
Comment (lonng) [inline pkg/cliutil/prepare/prepare.go:235]: Do you have tested this PR? I think the `uniqueHosts := set.NewStringSet()` should be placed out of `IterInstance`.

File: pkg/cliutil/prepare/prepare.go [hunk 4]
```diff
@@ -248,6 +266,20 @@ func CheckClusterPortConflict(clusterSpec *spec.SpecManager, clusterName string,
 				port:     port,
 			})
 		}
+
+		uniqueHosts := set.NewStringSet()
+		if !uniqueHosts.Exist(inst.GetHost()) {
+			uniqueHosts.Insert(inst.GetHost())
```
Comment (lonng) [inline pkg/cliutil/prepare/prepare.go:272]: ditto

File: pkg/cliutil/prepare/prepare.go (hunk 3 above)
Comment (AstroProfundis) [reply pkg/cliutil/prepare/prepare.go:235]: It should be in the loop, it "unique" for on cluster, not for all clusters

Comment (lonng) [reply pkg/cliutil/prepare/prepare.go:235]: Yes, it should be placed in the `for` loop, but above the L234.

File: pkg/utils/retry.go [hunk 5]
```diff
@@ -13,16 +13,99 @@
...
+		default:
+			time.Sleep(cfg.Delay)
+		}
+	}
+
+	return fmt.Errorf("operation exceeds the max retry attempts of %d", cfg.Attempts)
+}
+
+// IsTimeoutOrMaxRetry return true if it's timeout or reach max retry.
+func IsTimeoutOrMaxRetry(err error) bool {
```
Comment (lucklove) [inline pkg/utils/retry.go:98]: It's better to split into two functions: `IsTimeout` and `IsMaxRetry`. They have different semantics.

**Comment by lucklove** [comment]
Rest LGTM

**Pull request merged**